### ISSUE TITLE
Jest + Axios + flushPromises is not enough to wait for Mirage response

### ISSUE THREAD

**Issue by MartinMalinda** [miragejs/miragejs#458]
This might not be Mirage related, in that case I'm sorry.

In a Nuxt (Vue) app, I have a component test like this:

```
wrapper = render(MyForm);
fillIn(wrapper.find('input').element, 'valid@gmail.com');
wrapper.find('button').trigger('submit');
await flushPromises();
expect(onSuccess).toHaveBeenCalled();
```

That test was fine until I moved from `axios` mocking towards using Mirage. It seems like `await flushPromises();` is not enough anymore and the `expect(onSuccess).toHaveBeenCalled();` gets called too early.

In some scenarios I could introduce different waiting strategy. I could wait for error or success message rendering in the DOM (`await waitForElement(() =>  {}`); But in this case, there's the prop callback being called and that's it.

I can see that the route handler is reached before the test ends - even before the assertion (🙏 ) so all I need is a tiny bit of extra waiting, so it would be awesome if I could do something like

`await server.idle()`

There's one reliable solution and that's 

```
await waitFor(() => expect(onRegister).toHaveBeenCalled());
```

But it seems its significiantly slower (up to extra 50 - 100ms from my measurements, which makes sense, given it loops in 50ms intervals). With the `waitFor(() => {})`. This test takes up to 200ms which seems like it's more than it should. But I could be wrong here and it's normal:)

**Comment by dv297** [comment]
I also just hit this issue in my own Vue tests, where I was trying to use MirageJS, flush-promises, and @testing-library/vue. If I have time this weekend, I'd be happy to try and create a simplified recreation if it would help maintainers. As part of exploration, instead of using flushPromises, I used a setTimeout to add a delay before doing my assertions. My tests would fail unless I added ~500ms of delay using that method.

**Comment by dv297** [comment]
Hoping this helps, created a small sample project that shows a List component with a button that you can click to make the Axios call to fetch the data.

https://github.com/dv297/miragejs-axios-jest-recreation

The tests fail because the MirageJS response doesn't come back, despite the flush-promise call.

**Comment by MartinMalinda** [comment]
@dv297 hi Daniel I think in your case the long waiting is needed because you have environment `testing` not `test` and therefore the server has 400ms of waiting. Try changing the environment to `test` or set `timing` to 0. I believe flush promises won't be enough waiting in that case either though.

**Comment by dv297** [comment]
Ah, interesting. Never realized MirageJS had that behavior, but found it on the Workflow Tips page after you mentioned it. Pushed the changes, and yea, flush promises is still not enough. But can remove the need for such a high delay. Hopefully this is still a helpful reproduction.

**Comment by ryanto** [comment]
@MartinMalinda Yah, I remember running into this when testing Vue apps. Have you tried to configure `waitFor`'s timeout to get faster than 50ms? I think that's the best answer I can offer.

I recall `flushPromises` not being a good fit here because there's a couple of promises that the test ends up needing to wait on. I'm not really in love with the idea of await the next runloop, it feels similar to `setTimeout(..., 1)`.

Also `server.idle()` seems like a great idea, and it's come up in the past 👍. I can't offer an ETA on when we'd be able to get that API into mirage since we're currently focused on building a tutorial, fixing bugs, and shipping 1.0. Since it's a new API it seems like a post 1.0 thing.

I'd love to know if a 10ms waitFor helps at all.

**Comment by ryanto** [comment]
Whoops, I meant interval (not timeout)! Here's the doc... https://testing-library.com/docs/dom-testing-library/api-async#waitfor

**Comment by ryanto** [comment]
@dv297 To get your application tests working the best approach is to not wait a specific amount of time, but to wait for an element to be in the document. Check out the [guides on appearance and disappearance](https://testing-library.com/docs/guide-disappearance).

This should be enough to get your tests passing.

```diff
+ import { render, fireEvent, waitFor } from "@testing-library/vue";

// ...

  it("shows list elements from api", async () => {
    const { findByText, getByText, queryByText } = render(List);

    const retrieveButton = getByText("Retrieve List");
    await fireEvent.click(retrieveButton);

+    await waitFor(() => {
      expect(queryByText("ABC")).toBeInTheDocument();
      expect(queryByText("DEF")).toBeInTheDocument();
+    })
  });
```

I think this also has the benefit of making your tests more robust since your no longer relying on implementation details like time or currently resolved promises.

**Comment by mcoqblin-hiptest** [comment]
Hello,

I am bumping this issue because I ran into the same problem earlier this year when trying to setup Mirage in a test environment. I found the source of the problem and the solution, which involves no delays in tests 😄.

**tl;dr** Setting `server.timing` to **false** (not 0) **after** the server creation will allow you to simply use `await flushPromises` / `await act` / `await waitFor` without any delay.

### Reason

Mirage uses Pretender to handle routes, and includes the timing property.

To introduce that delay, Pretender uses `setTimeout`. This is why awaiting for promises has no effect here.

It is also not possible to use fake timers to instantly execute `setTimeout` calls, since for various feature reasons Pretender actually performs 50ms long `setTimeout` calls, and checks the actual elapsed time. It also does that using recursive calls, so attempting to fake `setTimeout` will result in a good old stack overflow.

### Solution

As a [documented feature](https://github.com/pretenderjs/pretender/blob/master/README.md), the `timing` parameter can also be a boolean. If set to false, it will make the request synchronous. After checking, I can confirm it skips every `setTimeout` calls.

Right now Mirage does not support having a false `timing` option when creating a server. Probably part of a bug by the way, as the line `this.timing = this.timing || config.timing || 400` can never result in `this.timing` being set to 0 in a development environment.

However it is possible to force it after the server creation, using `server.timing = false`. I suppose this is officially supported as the documentation mentions being able to force a timing for a specific test this way.

I can confirm it works wonderfully my test suite in a React environment using testing-library. For instance:

```
button = render(<Button>)

await act(async () => {
  fireEvent.click(button.container)
})

expect(onSuccess).toHaveBeenCalled()
```

It also works well with `waitFor` or `flushPromises`, and will not have any delay.

### Updating Mirage

I would suggest that instead of forcing `timing` to 0 for tests, it is forced to false instead. I don't think this would break existing tests, unless some nasty workaround has been done like waiting for `setTimeout` to be called.

For safety though, I also suggest either allowing `timing` to be false in config, and/or adding a boolean `async` option to be able to have more control over that behavior.

I am currently trying the changes on my side and will create a pull request when it is done.
